features/370-auth-ui/design.md

```diff
@@ -0,0 +1,393 @@
+All pages must be function properly even without JavaScript.
+
+### Customization of page templates
+
+The developer can customize the pages at 2 levels.
+They can keep using the builtin templates and customize the theme.
+Or they can replace the builtin ones with their own ones.
+
+#### Customizing the theme with CSS
+
+By studying the UI design of Auth UI, a few class names can generalized:
+
+- `.primary-txt`: The primary text.
+- `.secondary-txt`: The secondary text.
+- `.primary-btn`: The primary button.
+- `.secondary-btn`: The secondary button.
+- `.anchor`: The anchor.
+- `.text-input`: The text input.
+
+There are a few CSS pseudo classes that are applicable to the UI elements.
+
+- `:link`: Unvisited links.
+- `:visited`: Visited links.
+- `:hover`: Mouse hover state.
+- `:active`: Button being clicked.
+
+Buttons also have two variants.
+The primary button has background color while the secondary one has border.
+
+So the following CSS properties should be able to customize.
+
+- `color`
+- `background-color`
+- `border*`
+
+To customize all combinations, the developer can provide the CSS directly.
+The developer is strongly recommended to write CSS covering only
+the above mentioned classes, pseudo classes and properties.
+There is no validation to enforce this restriction.
+
+#### Customizing with own templates
+
+The developer can also provide their own templates via [the existing template mechanism](../251-templates/index.md).
+
+## The authorize endpoint
+
+The authorize endpoint `/_auth/_/authorize` allows the user to authenticate with configured authentication methods. It handles MFA challenge during authentication. If the user does not have an account yet, the user can sign up directly.
+
+### The authorization page
+
+The authorization page is the landing page of the authorize endpoint. It lists out the configured IdPs. It shows a text field for login ID. The login ID field is either a plain text input or a phone number input, depending on the type of the first login ID key.
+
+```
+|---------------------------|
+| Login with Google         |
+|---------------------------|
+| Login with Facebook       |
+|---------------------------|
+
+              Or
+
+|--------------------------------------|  |----------|
+| Enter in your email or username here |  | Continue |
+|--------------------------------------|  |----------|
+
+Login with a phone number instead.
+```
+
+If the user selects an IdP, the user is redirected to the authorization endpoint of that IdP. If the user continues with a login ID, the user is directed to the password page.
+
+Separating the login ID and the password into two pages allow future improvements, such as [Open ID Provider Issuer Discovery](https://openid.net/specs/openid-connect-discovery-1_0.html#IssuerDiscovery). For example, the user can type in their Google account name and then be redirected to the authorization endpoint of Google.
```

Comment:
A side note... have we created the feature issue for the things we have discussed / our Lxxx clients need? Which is to force loginID to use SSO if configured...? And maybe we should mention that here.

Reply:
That feature is independent of Auth UI. I will create another feature issue for that.

Reply:
Thanks yes (sort of independent, but after we implemented that feature we will have some code change on AuthUI to handle the redirect?)

Reply:
https://github.com/SkygearIO/features/issues/412

Reply:
> (sort of independent, but after we implemented that feature we will have some code change on AuthUI to handle the redirect?)

Edited the feature issue to include this suggestion.